// @lit-protocol/lit-auth-client/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAuthIdByAuthMethod = exports.getProviderFromUrl = exports.isSignInRedirect = exports.WebAuthnProvider = exports.StytchAuthFactorOtpProvider = exports.StytchOtpProvider = exports.LitRelay = exports.GoogleProvider = exports.EthWalletProvider = exports.DiscordProvider = exports.BaseProvider = exports.AppleProvider = void 0;
const tslib_1 = require("tslib");
const AppleProvider_1 = tslib_1.__importDefault(require("./lib/providers/AppleProvider"));
exports.AppleProvider = AppleProvider_1.default;
const BaseProvider_1 = require("./lib/providers/BaseProvider");
Object.defineProperty(exports, "BaseProvider", { enumerable: true, get: function () { return BaseProvider_1.BaseProvider; } });
const DiscordProvider_1 = tslib_1.__importDefault(require("./lib/providers/DiscordProvider"));
exports.DiscordProvider = DiscordProvider_1.default;
const EthWalletProvider_1 = tslib_1.__importDefault(require("./lib/providers/EthWalletProvider"));
exports.EthWalletProvider = EthWalletProvider_1.default;
const GoogleProvider_1 = tslib_1.__importDefault(require("./lib/providers/GoogleProvider"));
exports.GoogleProvider = GoogleProvider_1.default;
const StytchAuthFactorOtp_1 = tslib_1.__importDefault(require("./lib/providers/StytchAuthFactorOtp"));
exports.StytchAuthFactorOtpProvider = StytchAuthFactorOtp_1.default;
const StytchOtpProvider_1 = require("./lib/providers/StytchOtpProvider");
Object.defineProperty(exports, "StytchOtpProvider", { enumerable: true, get: function () { return StytchOtpProvider_1.StytchOtpProvider; } });
const WebAuthnProvider_1 = tslib_1.__importDefault(require("./lib/providers/WebAuthnProvider"));
exports.WebAuthnProvider = WebAuthnProvider_1.default;
const relay_1 = require("./lib/relay");
Object.defineProperty(exports, "LitRelay", { enumerable: true, get: function () { return relay_1.LitRelay; } });
const utils_1 = require("./lib/utils");
Object.defineProperty(exports, "isSignInRedirect", { enumerable: true, get: function () { return utils_1.isSignInRedirect; } });
Object.defineProperty(exports, "getProviderFromUrl", { enumerable: true, get: function () { return utils_1.getProviderFromUrl; } });
Object.defineProperty(exports, "getAuthIdByAuthMethod", { enumerable: true, get: function () { return utils_1.getAuthIdByAuthMethod; } });
