// @lit-protocol/types/src/lib/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthMethodType = exports.IRelayAuthStatus = void 0;
var IRelayAuthStatus;
(function (IRelayAuthStatus) {
    IRelayAuthStatus["InProgress"] = "InProgress";
    IRelayAuthStatus["Succeeded"] = "Succeeded";
    IRelayAuthStatus["Failed"] = "Failed";
})(IRelayAuthStatus = exports.IRelayAuthStatus || (exports.IRelayAuthStatus = {}));
/**
 * Supported PKP auth method types
 */
var AuthMethodType;
(function (AuthMethodType) {
    AuthMethodType[AuthMethodType["EthWallet"] = 1] = "EthWallet";
    AuthMethodType[AuthMethodType["LitAction"] = 2] = "LitAction";
    AuthMethodType[AuthMethodType["WebAuthn"] = 3] = "WebAuthn";
    AuthMethodType[AuthMethodType["Discord"] = 4] = "Discord";
    AuthMethodType[AuthMethodType["Google"] = 5] = "Google";
    AuthMethodType[AuthMethodType["GoogleJwt"] = 6] = "GoogleJwt";
    AuthMethodType[AuthMethodType["AppleJwt"] = 8] = "AppleJwt";
    AuthMethodType[AuthMethodType["StytchOtp"] = 9] = "StytchOtp";
    AuthMethodType[AuthMethodType["StytchEmailFactorOtp"] = 10] = "StytchEmailFactorOtp";
    AuthMethodType[AuthMethodType["StytchSmsFactorOtp"] = 11] = "StytchSmsFactorOtp";
    AuthMethodType[AuthMethodType["StytchWhatsAppFactorOtp"] = 12] = "StytchWhatsAppFactorOtp";
    AuthMethodType[AuthMethodType["StytchTotpFactorOtp"] = 13] = "StytchTotpFactorOtp";
})(AuthMethodType = exports.AuthMethodType || (exports.AuthMethodType = {}));
